Dart/src/com/jetbrains/lang/dart/analyzer/DartAnalysisServerService.java

```diff
@@ -225,9 +230,43 @@ public void run(@NotNull ProgressIndicator indicator) {
     @Override
     public void beforeDocumentChange(DocumentEvent e) {
       myServerData.onDocumentChanged(e);
+      synchronized (myLock) {
+        final Document document = e.getDocument();
+        myUpdatedDocuments.add(document);
```

Comment:
Hm...
You're right, we need more checks.
Checkin in `DartServerData` sounds good.